osu.Game/Rulesets/Edit/HitObjectComposer.cs

```diff
@@ -218,7 +218,7 @@ protected override void LoadComplete()
         /// <summary>
         /// Construct a relevant blueprint container. This will manage hitobject selection/placement input handling and display logic.
         /// </summary>
-        protected virtual ComposeBlueprintContainer CreateBlueprintContainer() => new ComposeBlueprintContainer(this);
+        protected virtual ComposeBlueprintContainer CreateBlueprintContainer() => new ComposeBlueprintContainer(this, Ruleset);
```

Comment:
To avoid breaking the ruleset API, `Ruleset` could be made `public` in `HitObjectComposer` and retrieved that way.

Here's a diff making that change:

```diff
diff --git a/osu.Game.Rulesets.Catch/Edit/CatchBlueprintContainer.cs b/osu.Game.Rulesets.Catch/Edit/CatchBlueprintContainer.cs
index da45e952ab..9408a9f95c 100644
--- a/osu.Game.Rulesets.Catch/Edit/CatchBlueprintContainer.cs
+++ b/osu.Game.Rulesets.Catch/Edit/CatchBlueprintContainer.cs
@@ -13,8 +13,8 @@ namespace osu.Game.Rulesets.Catch.Edit
 {
     public partial class CatchBlueprintContainer : ComposeBlueprintContainer
     {
-        public CatchBlueprintContainer(CatchHitObjectComposer composer, Ruleset ruleset)
-            : base(composer, ruleset)
+        public CatchBlueprintContainer(CatchHitObjectComposer composer)
+            : base(composer)
         {
         }
 
diff --git a/osu.Game.Rulesets.Catch/Edit/CatchHitObjectComposer.cs b/osu.Game.Rulesets.Catch/Edit/CatchHitObjectComposer.cs
index fec678a4b6..bbdffbf39c 100644
--- a/osu.Game.Rulesets.Catch/Edit/CatchHitObjectComposer.cs
+++ b/osu.Game.Rulesets.Catch/Edit/CatchHitObjectComposer.cs
@@ -148,7 +148,7 @@ public override SnapResult FindSnappedPositionAndTime(Vector2 screenSpacePositio
             return result;
         }
 
-        protected override ComposeBlueprintContainer CreateBlueprintContainer() => new CatchBlueprintContainer(this, Ruleset);
+        protected override ComposeBlueprintContainer CreateBlueprintContainer() => new CatchBlueprintContainer(this);
 
         [CanBeNull]
         private PalpableCatchHitObject getLastSnappableHitObject(double time)
diff --git a/osu.Game.Rulesets.Mania/Edit/ManiaBlueprintContainer.cs b/osu.Game.Rulesets.Mania/Edit/ManiaBlueprintContainer.cs
index 060073889d..05d8ccc73f 100644
--- a/osu.Game.Rulesets.Mania/Edit/ManiaBlueprintContainer.cs
+++ b/osu.Game.Rulesets.Mania/Edit/ManiaBlueprintContainer.cs
@@ -13,8 +13,8 @@ namespace osu.Game.Rulesets.Mania.Edit
 {
     public partial class ManiaBlueprintContainer : ComposeBlueprintContainer
     {
-        public ManiaBlueprintContainer(HitObjectComposer composer, Ruleset ruleset)
-            : base(composer, ruleset)
+        public ManiaBlueprintContainer(HitObjectComposer composer)
+            : base(composer)
         {
         }
 
diff --git a/osu.Game.Rulesets.Mania/Edit/ManiaHitObjectComposer.cs b/osu.Game.Rulesets.Mania/Edit/ManiaHitObjectComposer.cs
index ff59131c8a..5e577a2964 100644
--- a/osu.Game.Rulesets.Mania/Edit/ManiaHitObjectComposer.cs
+++ b/osu.Game.Rulesets.Mania/Edit/ManiaHitObjectComposer.cs
@@ -68,7 +68,7 @@ protected override DrawableRuleset<ManiaHitObject> CreateDrawableRuleset(Ruleset
         }
 
         protected override ComposeBlueprintContainer CreateBlueprintContainer()
-            => new ManiaBlueprintContainer(this, Ruleset);
+            => new ManiaBlueprintContainer(this);
 
         protected override IReadOnlyList<HitObjectCompositionTool> CompositionTools => new HitObjectCompositionTool[]
         {
diff --git a/osu.Game.Rulesets.Osu/Edit/OsuBlueprintContainer.cs b/osu.Game.Rulesets.Osu/Edit/OsuBlueprintContainer.cs
index 58bae7e2b2..173a664902 100644
--- a/osu.Game.Rulesets.Osu/Edit/OsuBlueprintContainer.cs
+++ b/osu.Game.Rulesets.Osu/Edit/OsuBlueprintContainer.cs
@@ -15,8 +15,8 @@ namespace osu.Game.Rulesets.Osu.Edit
 {
     public partial class OsuBlueprintContainer : ComposeBlueprintContainer
     {
-        public OsuBlueprintContainer(HitObjectComposer composer, Ruleset ruleset)
-            : base(composer, ruleset)
+        public OsuBlueprintContainer(HitObjectComposer composer)
+            : base(composer)
         {
         }
 
diff --git a/osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs b/osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs
index cf22059133..09ddc420a7 100644
--- a/osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs
+++ b/osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs
@@ -88,7 +88,7 @@ private void load()
         }
 
         protected override ComposeBlueprintContainer CreateBlueprintContainer()
-            => new OsuBlueprintContainer(this, Ruleset);
+            => new OsuBlueprintContainer(this);
 
         public override string ConvertSelectionToString()
             => string.Join(',', selectedHitObjects.Cast<OsuHitObject>().OrderBy(h => h.StartTime).Select(h => (h.IndexInCurrentCombo + 1).ToString()));
diff --git a/osu.Game.Rulesets.Taiko/Edit/TaikoBlueprintContainer.cs b/osu.Game.Rulesets.Taiko/Edit/TaikoBlueprintContainer.cs
index 0081f89b7c..6be22f3af0 100644
--- a/osu.Game.Rulesets.Taiko/Edit/TaikoBlueprintContainer.cs
+++ b/osu.Game.Rulesets.Taiko/Edit/TaikoBlueprintContainer.cs
@@ -12,8 +12,8 @@ namespace osu.Game.Rulesets.Taiko.Edit
 {
     public partial class TaikoBlueprintContainer : ComposeBlueprintContainer
     {
-        public TaikoBlueprintContainer(HitObjectComposer composer, Ruleset ruleset)
-            : base(composer, ruleset)
+        public TaikoBlueprintContainer(HitObjectComposer composer)
+            : base(composer)
         {
         }
 
diff --git a/osu.Game.Rulesets.Taiko/Edit/TaikoHitObjectComposer.cs b/osu.Game.Rulesets.Taiko/Edit/TaikoHitObjectComposer.cs
index 4109bc4887..cff5731181 100644
--- a/osu.Game.Rulesets.Taiko/Edit/TaikoHitObjectComposer.cs
+++ b/osu.Game.Rulesets.Taiko/Edit/TaikoHitObjectComposer.cs
@@ -24,6 +24,6 @@ public TaikoHitObjectComposer(TaikoRuleset ruleset)
         };
 
         protected override ComposeBlueprintContainer CreateBlueprintContainer()
-            => new TaikoBlueprintContainer(this, Ruleset);
+            => new TaikoBlueprintContainer(this);
     }
 }
diff --git a/osu.Game/Rulesets/Edit/HitObjectComposer.cs b/osu.Game/Rulesets/Edit/HitObjectComposer.cs
index 8279150aa8..b5b7400f64 100644
--- a/osu.Game/Rulesets/Edit/HitObjectComposer.cs
+++ b/osu.Game/Rulesets/Edit/HitObjectComposer.cs
@@ -45,8 +45,6 @@ public abstract partial class HitObjectComposer<TObject> : HitObjectComposer, IP
     {
         protected IRulesetConfigManager Config { get; private set; }
 
-        protected readonly Ruleset Ruleset;
-
         // Provides `Playfield`
         private DependencyContainer dependencies;
 
@@ -74,8 +72,8 @@ public abstract partial class HitObjectComposer<TObject> : HitObjectComposer, IP
         private IBindable<bool> hasTiming;
 
         protected HitObjectComposer(Ruleset ruleset)
+            : base(ruleset)
         {
-            Ruleset = ruleset;
         }
 
         protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnlyDependencyContainer parent) =>
@@ -218,7 +216,7 @@ protected override void LoadComplete()
         /// <summary>
         /// Construct a relevant blueprint container. This will manage hitobject selection/placement input handling and display logic.
         /// </summary>
-        protected virtual ComposeBlueprintContainer CreateBlueprintContainer() => new ComposeBlueprintContainer(this, Ruleset);
+        protected virtual ComposeBlueprintContainer CreateBlueprintContainer() => new ComposeBlueprintContainer(this);
 
         /// <summary>
         /// Construct a drawable ruleset for the provided ruleset.
@@ -419,8 +417,11 @@ public override SnapResult FindSnappedPositionAndTime(Vector2 screenSpacePositio
     [Cached]
     public abstract partial class HitObjectComposer : CompositeDrawable, IPositionSnapProvider
     {
-        protected HitObjectComposer()
+        public readonly Ruleset Ruleset;
+
+        protected HitObjectComposer(Ruleset ruleset)
         {
+            Ruleset = ruleset;
             RelativeSizeAxes = Axes.Both;
         }
 
diff --git a/osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs b/osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs
index fc23f1d1b4..3c9c6094e2 100644
--- a/osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs
@@ -47,10 +47,10 @@ public partial class ComposeBlueprintContainer : EditorBlueprintContainer
         /// </remarks>
         public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => true;
 
-        public ComposeBlueprintContainer(HitObjectComposer composer, Ruleset ruleset)
+        public ComposeBlueprintContainer(HitObjectComposer composer)
             : base(composer)
         {
-            this.ruleset = ruleset;
+            ruleset = composer.Ruleset;
 
             placementBlueprintContainer = new Container<PlacementBlueprint>
             {

```